#### CSCE 670 :: Information Storage and Retrieval :: Texas A&M University :: Spring 2017


# Homework 3:  Classification Cook-off: Naive Bayes vs Rocchio (plus a little bit of recommenders)

### 100 points [5% of your final grade]

### Due: Wednesday, March 29 by 11:59pm

*Goals of this homework:* Hands-on practice building and evaluating classifiers.

*Submission Instructions:* To submit your homework, rename this notebook as YOUR_UIN_hw3.ipynb. Submit this notebook via eCampus. Your notebook should be completely self-contained, with the results visible in the notebook. 

*Late submission policy:* For this homework, you may use up to three of your late days, meaning that no submissions will be accepted after Saturday April 1 at 11:59pm.

# Part 0: Yelp review data

In this assignment, given a Yelp review, your task is to implement two classifiers to predict if the business category of this review is "food-relevant" or not, **only based on the review text**. The data is from the [Yelp Dataset Challenge](https://www.yelp.com/dataset_challenge).

## Build the training data

First, you will need to download this data file as your training data: [training_data.json](https://drive.google.com/open?id=0B_13wIEAmbQMdzBVTndwenoxQlk) 

The training data file includes 40,000 Yelp reviews. Each line is a json-encoded review, and **you should only focus on the "text" field**. As same as in homework 1, you should tokenize the review text by using the regular expression "\W+" (we discussed it in [this Piazza post](https://piazza.com/class/ixkk1fy863r1vs?cid=29). Do NOT remove stop words. **Do casefolding but no stemming**.

The label (class) information of each review is in the "label" field. It is **either "Food-relevant" or "Food-irrelevant"**.

## Testing data

We provide 100 yelp reviews here: [testing_data.json](https://drive.google.com/open?id=0B_13wIEAmbQMbXdyTkhrZDN4Wms). The testing data file has the same format as the training data file. Again, you can get the label informaiton in the "label" field. Only use it when you evalute your classifiers.

# Part 1: Naive Bayes classifier [35 points]

In this part, you will implement a Naive Bayes classifier, which outputs the probabilities that a given review belongs to each class.

Use a mixture model that mixes the probability from the document with the general collection frequency of the word. **You should use lambda = 0.7**. Be careful about the decimal rounding since multiplying many probabilities can generate a tiny value. We will not grade on the exact probability value, so feel free to change to logorithm summation (it's not required, though). If the tie case happens, **always go to the "Food-irrelevant" side**.

### What to report

* For the entire testing dataset, report the overall accuracy.
* For the class "Food-relevant", report the precision and recall.
* For the class "Food-irrelevant", report the precision and recall.

We will also grade on the quality of your code. So make sure that your code is clear and readable.

In [6]:
# Build the naive bayes classifier
# Insert as many cells as you want
import json
import math
import re
import sys
from collections import defaultdict

lamda = 0.7
klassesCount = defaultdict(lambda: 0.0)
totalDocs = 0
vocab = set()
klassTotalWordCount = defaultdict(lambda: 0.0)
klassWordCount = defaultdict(lambda: 0.0)
klasses = ["Food-relevant", "Food-irrelevant"]

def add_review(review, klass):
    global klassesCount, totalDocs, vocab, klassTotalWordCount, klassWordCount
    totalDocs += 1
    klassesCount[klass] += 1
    #review = review.lower()
    temp_list = re.split('\W+', review)
    temp_list = filter(None, temp_list)
    for word in temp_list:
        word = word.lower()
        if word == "":
            continue
        vocab.add(word)
        klassTotalWordCount[klass] += 1
        klassWordCount[(klass, word)] += 1

def parse_json():
    with open('training_data.json') as data_file:
        for line in data_file:
            data = json.loads(line)
            klass = data["label"]
            review = data["text"]
            add_review(review, klass)

parse_json()
#print "vocab size : ", len(vocab)
#print "relevant tokens : ", klassTotalWordCount[klasses[0]], "  irrelevant : ", klassTotalWordCount[klasses[1]]
            
def nb_classify(review):
    klass_prob = {}
    review = review.lower()
    temp_list = re.split('\W+', review)
    temp_list = filter(None, temp_list)
    for klass in klasses:
        klass_prob[klass] = math.log10(float(klassesCount[klass]) / totalDocs)
        for word in temp_list:
            if word == "" or word not in vocab:
                continue
            #word = word.lower()
            num2 = num1 = klassWordCount[(klass, word)]
            den2 = den1 = klassTotalWordCount[klass]
            if klass == klasses[0]:
                num2 += klassWordCount[(klasses[1], word)]
                den2 += klassTotalWordCount[klasses[1]]
            else:
                num2 += klassWordCount[(klasses[0], word)]
                den2 += klassTotalWordCount[klasses[0]]
            klass_prob[klass] += math.log10(lamda*(float(num1) / den1) + (1 - lamda)*(float(num2) / den2))
    return klass_prob

In [7]:
# Apply your classifier on the test data. Report the results.
# Insert as many cells as you want
def test():
    confusion_matrix = defaultdict(lambda: defaultdict(lambda: 0.0))
    with open('testing_data.json') as data_file:
        total_docs = 0.0
        correct_docs = 0.0
        incorrect_docs = {}
        for line in data_file:
            total_docs += 1
            data = json.loads(line)
            klass_prob = nb_classify(data["text"])
            predicted_class = None
            max_prob = -sys.maxint -1
            for klass in klasses:
                temp_prob = klass_prob[klass]
                if temp_prob >= max_prob:
                    max_prob = temp_prob
                    predicted_class = klass
            #print klass_prob[klasses[0]], " ", klass_prob[klasses[1]], " ", predicted_class, " ", data["label"]
            if predicted_class == data["label"]:
                correct_docs += 1
                confusion_matrix[predicted_class]["True"] += 1
            else:
                incorrect_docs[total_docs] = (data["text"], data["label"])
                confusion_matrix[predicted_class]["False"] += 1
    print "accuracy : ", correct_docs / total_docs
    print klasses[0], " : "
    print "Precision : ", confusion_matrix[klasses[0]]["True"] / (confusion_matrix[klasses[0]]["True"] + confusion_matrix[klasses[0]]["False"])
    print "Recall : ", confusion_matrix[klasses[0]]["True"] / (confusion_matrix[klasses[0]]["True"] + confusion_matrix[klasses[1]]["False"])
    
    print klasses[1], " : "
    print "Precision : ", confusion_matrix[klasses[1]]["True"] / (confusion_matrix[klasses[1]]["True"] + confusion_matrix[klasses[1]]["False"])
    print "Recall : ", confusion_matrix[klasses[1]]["True"] / (confusion_matrix[klasses[1]]["True"] + confusion_matrix[klasses[0]]["False"])
    '''
    for key in incorrect_docs:
        #print key, " ", incorrect_docs[key]
        print key
    '''
test()


accuracy :  0.89
Food-relevant  : 
Precision :  0.901408450704
Recall :  0.941176470588
Food-irrelevant  : 
Precision :  0.862068965517
Recall :  0.78125


# Part 2: Rocchio classifier [35 points]

In this part, your job is to implement a Rocchio classifier for "food-relevant vs. food-irrelevant". You need to aggregate all the reviews of each class, and find the center. **Use the normalized raw term frequency**.


### What to report

* For the entire testing dataset, report the overall accuracy.
* For the class "Food-relevant", report the precision and recall.
* For the class "Food-irrelevant", report the precision and recall.

We will also grade on the quality of your code. So make sure that your code is clear and readable.

In [8]:
def magnitude(u):
    return math.sqrt(sum(u[i]*u[i] for i in range(len(u))))

def magnitude_map(u):
    return math.sqrt(sum(u[i]*u[i] for i in u))

def normalize(u):
    mag = magnitude(u)
    if mag == 0:
        return u
    return [float(u[i])/mag for i in range(len(u))]

def normalize_map(u):
    mag = magnitude_map(u)
    if mag == 0:
        return u
    for i in u:
        u[i] = float(u[i])/mag
    return u

def centroid(u, num_docs):
    return [float(u[i])/num_docs for i in range(len(u))]

In [ ]:
# Build the Rocchio classifier
# Insert as many cells as you want
klasses = ["Food-relevant", "Food-irrelevant"]
klass_vec = defaultdict(lambda: defaultdict(lambda: 0.0))
doc_count = defaultdict(lambda: 0.0)
vocab = set()

def tokenize(review):
    review = review.lower()
    word_list = re.split('\W+', review)
    word_list = filter(None, word_list)
    temp_vec = defaultdict(lambda: 0.0)
    for word in word_list:
        #word = word.lower()
        vocab.add(word)
        temp_vec[word] += 1
    return normalize_map(temp_vec)

def rocchio_training():
    with open("training_data.json") as data_file:
        for line in data_file:
            data = json.loads(line)
            label = data["label"]
            doc_count[label] += 1
            review = data["text"]
            doc_vec = tokenize(review)
            for word in doc_vec:
                klass_vec[label][word] += doc_vec[word]

def gen_centroid_vecs():
    for klass in klasses:
        num_docs = doc_count[klass]
        for word in vocab:
            klass_vec[klass][word] = float(klass_vec[klass][word]) / num_docs
            
def manhattan_dist(klass, doc_vec):
    dist = 0.0
    for word in vocab:
        dist += abs(klass_vec[klass][word] - doc_vec[word])
    return dist

def euclidean_dist(klass, doc_vec):
    dist = 0.0
    for word in vocab:
        dist += ((klass_vec[klass][word] - doc_vec[word])*(klass_vec[klass][word] - doc_vec[word]))
    
    return math.sqrt(dist)

def rocchio_classify(review):
    doc_vec = tokenize(review)
    min_dist = sys.maxint
    predicted_klass = None
    for klass in klasses:
        '''
        dist = 0.0
        for word in vocab:
            dist += abs(klass_vec[klass][word] - doc_vec[word])
        '''
        #dist = manhattan_dist(klass, doc_vec)
        dist = euclidean_dist(klass, doc_vec)
        #print dist
        if dist <= min_dist:
            min_dist = dist
            predicted_klass = klass
    return predicted_klass


rocchio_training()
gen_centroid_vecs()

In [ ]:
# Apply your classifier on the test data. Report the results.
# Insert as many cells as you want
def test():
    incorrect_docs = {}
    total_docs = 0.0
    confusion_matrix = defaultdict(lambda: defaultdict(lambda: 0.0))
    with open("testing_data.json") as data_file:
        for line in data_file:
            total_docs += 1.0
            data = json.loads(line)
            review = data["text"]
            predicted_klass = rocchio_classify(review)
            if predicted_klass != data["label"]:
                incorrect_docs[total_docs] = (data["text"], data["label"])
                confusion_matrix[predicted_klass]["False"] += 1
            else:
                confusion_matrix[predicted_klass]["True"] += 1
    
    print "accuracy : ", float(total_docs - len(incorrect_docs)) / total_docs
    print klasses[0], " : "
    print "Precision : ", float(confusion_matrix[klasses[0]]["True"]) / (confusion_matrix[klasses[0]]["True"] + confusion_matrix[klasses[0]]["False"])
    print "Recall : ", float(confusion_matrix[klasses[0]]["True"]) / (confusion_matrix[klasses[0]]["True"] + confusion_matrix[klasses[1]]["False"])
                                                                                       
    print klasses[1], " : "
    print "Precision : ", float(confusion_matrix[klasses[1]]["True"]) / (confusion_matrix[klasses[1]]["True"] + confusion_matrix[klasses[1]]["False"])
    print "Recall : ", float(confusion_matrix[klasses[1]]["True"]) / (confusion_matrix[klasses[1]]["True"] + confusion_matrix[klasses[0]]["False"])  
    '''
    #print len(incorrect_docs), " ", total_docs
    for key in incorrect_docs:
        #print key, " : ", incorrect_docs[key]
        print key
    '''
test()

# Part 3: Naive Bayes vs. Rocchio [20 points]

Which method gives the better results? In terms of what? How did you compare them? Can you explain why you observe what you do? Write 1-3 paragraphs below.

**Add your answer here:**

Here, Naive Bayes gives out better results than Rocchio (for all accuracy, precision and recall) as shown by results below. 

NAIVE BAYES  - 

accuracy :  0.89

Food-relevant  : 

Precision :  0.901408450704

Recall :  0.941176470588

Food-irrelevant  : 

Precision :  0.862068965517

Recall :  0.78125

ROCCHIO - 

accuracy :  0.67

Food-relevant  : 

Precision :  0.818181818182

Recall :  0.661764705882

Food-irrelevant  : 

Precision :  0.488888888889

Recall :  0.6875

Naive bayes is a language model(generative model), whereas Rocchio is a classification algorithm. Rochio can sometimes lead to bad(confused) results. This is because, in Rocchio, we calculate the centroids of the documents in each class. If the distribution of the documents in a class is not consistent, like the docs are distributed in two clusters far away, then the centroid will lie in between. Thus, it may happen that the centroid of one class may be lying close to documents of other class. And thus, classification will be wrong here. 

Naive bayes is a probabilistic model which is immune from this problem. Thus naive bayes is quite robust as compared to Rocchio. Rocchio is preferred when we know the distribution of documents belonging to different classes in advance. Naive Bayes is not affected by the distribution of the data. Also, Rocchio is not preferred in case of multimodal distribution

Also, I observed that the Rocchio results are getting affected by the way I calculate the distance(manhattan/eucledian) or the way I normalize the vector. There is no such variation in Naive Bayes.
For Rocchio, accuracy also depends on the fact that how we normalize and represent the doc vectors.

# Part 4: Recommenders [10 points]

Finally, since we've begun our discussion of recommenders, let's do a quick problem too:

The table below is a utility matrix, representing the ratings, on a 1–5 star scale, of eight items, *a* through *h*, by three users *A*, *B*, and *C*. 
<pre>

  | a  b  c  d  e  f  g  h
--|-----------------------
A | 4  5     5  1     3  2
B |    3  4  3  1  2  1
C | 2     1  3     4  5  3

</pre>

Compute the following from the data of this matrix.

(a) Treating the utility matrix as boolean, compute the Jaccard distance between each pair of users.

(b) Repeat Part (a), but use the cosine distance.

(c) Treat ratings of 3, 4, and 5 as 1 and 1, 2, and blank as 0. Compute the Jaccard distance between each pair of users.

(d) Repeat Part (c), but use the cosine distance.

(e) Normalize the matrix by subtracting from each nonblank entry the average
value for its user.

(f) Using the normalized matrix from Part (e), compute the cosine distance
between each pair of users.

(g) Which of the approaches above seems most reasonable to you? Give a one or two sentence argument supporting your choice.

**Add your answer here:**

(a)A-B = 1/2, B-C = 1/2, A-C = 1/2

(b)A-B = 2/3, B-C = 2/3, A-C = 2/3

(c)A-B = 2/5, B-C = 1/6, A-C = 2/6

(d)A-B = 1/1.73 or 1/sqrt(3), B-C = 1/3.46 or 1/(2*sqrt(3)), A-C = 1/2

(e)
A - avg = 3.33
B - avg = 2.33
C - avg = 3.0
<pre>

  |  a    b    c    d     e     f     g    h
--|--------------------------------------------
A | 0.67 1.67      1.67 -2.33       -0.33 -1.33
B |      0.67 1.67 0.67 -1.33 -0.33 -1.33
C |-1.0      -2.0  0.0         1.0   2.0   0.0

</pre>

(f)cos(A-B) = (1.67*0.67 + 1.67*0.67 + -2.33*-1.33 + -0.33*-1.33) / (sqrt(0.67*0.67 + 1.67*1.67 + 1.67*1.67 + -2.33*-2.33 + -0.33*-0.33 + -1.33*-1.33) * sqrt(0.67*0.67 + 1.67*1.67 + 0.67*0.67 + -1.33*-1.33 + -0.33*-0.33 + -1.33*-1.33)) = 0.58408219372

cos(B-C) = (1.67*-2.0 + -0.33*1.0 + -1.33*2.0) / (sqrt(0.67*0.67 + 1.67*1.67 + 0.67*0.67 + -1.33*-1.33 + -0.33*-0.33 + -1.33*-1.33) * sqrt(-1.0*-1.0 + -2.0*-2.0 + 1.0*1.0 + 2.0*2.0)) = -0.73918138757

cos(A-C) = (0.67*-1.0 + -0.33*2.0) / (sqrt(0.67*0.67 + 1.67*1.67 + 1.67*1.67 + -2.33*-2.33 + -0.33*-0.33 + -1.33*-1.33) * sqrt(-1.0*-1.0 + -2.0*-2.0 + 1.0*1.0 + 2.0*2.0)) = -0.11518109075

(g) Cosine distances calculated from the normalized matrix make most sense to me(last approach). This is followed by approach in part d and approach in part c. 
    - Here, as seen from the results, A-B are most similar followed by A-C and B-C. This is also evident if we eyeball the matrix.
    - Normalizing the matrix introduces consistency, sanity in the ratings. This can be seen as if a user ,on an average, rates movies highly while other user berates movies highly. This brings them to the same level.
    - cosine similarity takes absolute values in account instead of the boolean values like in Jaccard. So, in Jaccard, we lose some of the critical info which gets preserved and accounted for in Cosine similarity.
    - In the third approach, we considered the normalized values, which resulted into most reasonable similarity values.